spock-core/src/main/java/org/spockframework/runtime/BaseSpecRunner.java

```diff
@@ -176,28 +174,41 @@ public Object invoke(Object target, Object... arguments) {
   public void doRunSetupSpec(SpecInfo spec) {
     runSetupSpec(spec.getSuperSpec());
     for (MethodInfo method : spec.getSetupSpecMethods()) {
-      if (runStatus != OK) return;
-      invoke(currentInstance, method);
+      if (runStatus.get() != OK) return;
+      invoke(NO_CURRENT_FEATURE, NO_CURRENT_ITERATION, sharedInstance, sharedInstance, method);
     }
   }
 
   private void runFeatures() {
-    for (FeatureInfo feature : spec.getAllFeaturesInExecutionOrder()) {
+    List<FeatureInfo> allFeaturesInExecutionOrder = spec.getAllFeaturesInExecutionOrder();
+    Scheduler scheduler = this.scheduler.deriveScheduler(!spec.isSupportParallelExecution());
+    for (int i = 0; i < allFeaturesInExecutionOrder.size(); i++) {
+      final FeatureInfo feature = allFeaturesInExecutionOrder.get(i);
       if (resetStatus(FEATURE) != OK) return;
-      currentFeature = feature;
-      runFeature();
-      currentFeature = null;
+
+      final boolean isLastFeature = i != allFeaturesInExecutionOrder.size() - 1;
+      if (isLastFeature) {
```

Comment:
The variable is named wrong, from the expression it should be `isNotLastFeature`. Maybe it would be better to write it like this.

```
boolean isLastFeature = i == allFeaturesInExecutionOrder.size() - 1;
if (isLastFeature) {// let's not waste threads and execute last iteration in current
  runFeature(feature);
} else {
  scheduler.schedule(new Runnable() {
...
```